#! /bin/sh -e

# All lines beginning with `# DP:' are a description of the patch.
# DP: Description: @DESCRIPTION@
# DP: Related bugs: 
# DP: Dpatch author: 
# DP: Patch author: 
# DP: Upstream status: [In CVS | Debian-Specific | Pending | Not submitted ]
# DP: Status Details: 

PATCHLEVEL=1

if [ $# -ne 2 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch) patch -d "$2" -f --no-backup-if-mismatch -p$PATCHLEVEL < $0;;
    -unpatch) patch -d "$2" -f --no-backup-if-mismatch -R -p$PATCHLEVEL < $0;;
    *)
	echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
	exit 1
esac
exit 0

# append the patch here and adjust the -p? flag in the patch calls.
diff -urN src.orig/nscd/grpcache.c src/nscd/grpcache.c
--- src.orig/nscd/grpcache.c	2005-07-19 05:23:49.000000000 +0200
+++ src/nscd/grpcache.c	2005-12-17 21:04:44.000000000 +0100
@@ -33,6 +33,7 @@
 #include <unistd.h>
 #include <sys/mman.h>
 #include <stackinfo.h>
+#include <limits.h>
 
 #include "nscd.h"
 #include "dbg_log.h"
diff -urN src.orig/nscd/pwdcache.c src/nscd/pwdcache.c
--- src.orig/nscd/pwdcache.c	2005-02-26 02:24:11.000000000 +0100
+++ src/nscd/pwdcache.c	2005-12-17 21:04:26.000000000 +0100
@@ -33,6 +33,7 @@
 #include <unistd.h>
 #include <sys/mman.h>
 #include <stackinfo.h>
+#include <limits.h>
 
 #include "nscd.h"
 #include "dbg_log.h"
diff -urN src.orig/nscd/mem.c src/nscd/mem.c
--- src.orig/nscd/mem.c	2004-09-13 07:56:03.000000000 +0200
+++ src/nscd/mem.c	2005-12-19 10:49:44.000000000 +0100
@@ -493,7 +493,8 @@
 	  size_t newtotal = (sizeof (struct database_pers_head)
 			     + db->head->module * sizeof (ref_t)
 			     + new_data_size);
-
+#define KFREEBSD_HACK 1
+#ifndef KFREEBSD_HACK
 	  if ((!db->mmap_used || ftruncate (db->wr_fd, newtotal) != 0)
 	      /* Try to resize the mapping.  Note: no MREMAP_MAYMOVE.  */
 	      && mremap (db->head, oldtotal, newtotal, 0) == 0)
@@ -502,6 +503,7 @@
 	      tried_resize = true;
 	      goto retry;
 	    }
+#endif	    
 	}
 
       if (! db->last_alloc_failed)
